ui/src/funding/contract.ts

```diff
@@ -87,59 +75,272 @@ export class PoolContract {
+        } catch {
+          // Ignore unmatching events
+          continue;
+        }
+        const amtPerSec = eventLog.amtPerSec;
+        if (amtPerSec.isZero()) {
+          // Ignore events stopping sending
+          continue;
+        }
+        totalAmtPerSec = totalAmtPerSec.add(amtPerSec);
+        receivers.push(eventLog.receiver);
+      }
 
-  async collectable(): Promise<Big> {
-    return this.contract.collectable().then(ethereum.toBaseUnit);
-  }
+      const balance = Big(event.args.balance.toString());
+      const timestamp = (await event.getBlock()).timestamp;
+      const amtPerSec = Big(totalAmtPerSec.toString());
+      const balanceFn = (now: Date) => {
+        // The Ethereum timestamps are in seconds
+        const timestampNow = Math.floor(now.getTime() / 1000);
+        // Block timestamp in the future
+        if (timestampNow < timestamp) return balance;
+        const spent = amtPerSec.mul(timestampNow - timestamp);
+        // Spent everything, return the unspendable reminder
+        if (spent > balance) return balance.mod(amtPerSec);
+        return balance.minus(spent);
+      };
+
+      const weeklyBudget = amountPerSecToWeeklyBudget(event.args.amtPerSec);
 
-  async weeklyBudget(): Promise<Big> {
-    return this.contract.getAmtPerSec().then(amountPerSecToWeeklyBudget);
+      return { balanceFn, weeklyBudget, receivers };
+    }
+
+    const filter = this.contract.filters.SenderUpdated(sender);
+
+    const listener = async (
+      _sender: unknown,
+      _balance: unknown,
+      _amtPerSec: unknown,
+      event: Event
+    ) => onUpdated(await getSenderData(event));
+    this.contract.on(filter, listener);
+
+    const lastEvents = await this.contract.queryFilter(filter);
+    const lastEvent = lastEvents.pop();
+    if (lastEvent) {
+      onUpdated(await getSenderData(lastEvent));
+    } else {
+      // Sender never configured
+      onUpdated({
+        balanceFn: () => Big(0),
+        weeklyBudget: Big(0),
+        receivers: [],
+      });
+    }
+
+    return () => this.contract.off(filter, listener);
   }
 
-  async receivers(): Promise<PoolReceiver[]> {
-    return this.contract.getAllReceivers();
+  // Start watching the state of a given pool receiver.
+  // `onUpdated` is called immediately with the latest state.
+  // Returns a function, which unwatches the state when called.
+  async watchPoolReceiver(
+    receiver: string,
+    onUpdated: (collectableFn: (now: Date) => Big) => void
+  ): Promise<() => void> {
+    const cycleSecs = (await this.contract.cycleSecs()).toNumber();
+
+    interface StreamFragment {
+      // Timestamp since which a stream sends funds (inclusive)
+      since: number;
+      // Timestamp to which a stream sends funds (exclusively)
+      to: number;
+      amtPerSec: BigNumber;
+    }
+    // For each sender stores a list of non-overlapping stream fragments
+    // in chronological order
+    const streams = new Map<string, StreamFragment[]>();
```

Comment:
This is state that is mutated by events through `onStreamUpdate` and read by `collectableFn`. I think it would make sense to encapsulate this in a class to better separate it from the rest of the code.
```typescript
class Collectable {
  private streams: Map<string, StreamFragment[]>;

  function collectableFn(now: Date): Big
  function onStreamUpdated(event: Event): Promise<void>
}
```

Reply:
Fixed